src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/Workspace.cs

```diff
@@ -279,6 +249,29 @@ async Task ProcessInitialEvaluationDataAsync(CancellationToken cancellationToken
                 // Let the exception escape, to unsubscribe data sources.
                 throw;
             }
+
+            (string languageName, string projectFilePath) GetProjectData()
+            {
+                snapshot.Properties.TryGetValue(ConfigurationGeneral.LanguageServiceNameProperty, out string? languageName);
+                snapshot.Properties.TryGetValue(ConfigurationGeneral.MSBuildProjectFullPathProperty, out string? projectFilePath);
+
+                return (string.IsNullOrEmpty(languageName), string.IsNullOrEmpty(projectFilePath)) switch
+                {
+                    (true, false) => throw Fail("Missing language name."),
+                    (false, true) => throw Fail("Missing project file path."),
+                    (true, true) => throw Fail("Missing language name and project file path"),
+                    (false, false) => (languageName, projectFilePath)
+                };
+
+                Exception Fail(string reason)
+                {
+                    // Insufficient data to initialize the language service.
+                    _state = WorkspaceState.Failed;
+
+                    // Throw, so we fault and tear down the dataflow.
+                    throw new($"Insufficient project data to initialize the language service. {reason}");
```

Comment:
Did you mean to throw the exception from here? This is throwing and not returning the exception to where you are also throwing in the switch expression above. So, the throw in the switch expression is only making the compiler happy since it'll never be hit at runtime.

Reply:
Yes this is intentional, but a good catch all the same! The reason is that each fork of the switch expression has to return a value, or throw. 

It might be possible to annotate this method with `[DoesNotReturn]` though I'm not sure whether that'll appease the compiler's control flow analysis here.

Regardless, the pattern used here is not uncommon. It's used in vs-validation for example, such as [here](https://github.com/microsoft/vs-validation/blob/3caa8c90b5940929b73862ac49dd8a514bdab96f/src/Microsoft.VisualStudio.Validation/Verify.cs#L80-L93) and [here](https://github.com/microsoft/vs-validation/blob/3caa8c90b5940929b73862ac49dd8a514bdab96f/src/Microsoft.VisualStudio.Validation/Requires.cs#L489-L520).